File: src/middleware/trafficControl.ts
import { MiddlewareHandler } from "hono";
import { cors } from "hono/cors";
import type { Bucket } from "../types/trafficControl";

const buckets = new Map<string, Bucket>();
const RATE_LIMIT_WINDOW_MS = Number(process.env.RATE_LIMIT_WINDOW_MS ?? 60_000);
const RATE_LIMIT_MAX = Number(process.env.RATE_LIMIT_MAX ?? 60);
const SLOW_DOWN_WINDOW_MS = Number(process.env.SLOW_DOWN_WINDOW_MS ?? 60_000);
const SLOW_DOWN_DELAY_AFTER = Number(process.env.SLOW_DOWN_DELAY_AFTER ?? 20);
const SLOW_DOWN_DELAY_MS = Number(process.env.SLOW_DOWN_DELAY_MS ?? 250);
const SLOW_DOWN_MAX_DELAY_MS = Number(process.env.SLOW_DOWN_MAX_DELAY_MS ?? 2_000);
const BUCKET_MAX_SIZE = Number(process.env.RATE_LIMIT_BUCKET_MAX_SIZE ?? 50_000);
const BUCKET_SWEEP_INTERVAL_MS = Number(process.env.RATE_LIMIT_SWEEP_INTERVAL_MS ?? 30_000);

function getClientKey(req: Request): string {
  const forwarded = req.headers.get("x-forwarded-for");
  if (forwarded) return forwarded.split(",")[0].trim();
  const realIp = req.headers.get("x-real-ip");
  if (realIp) return realIp;
  return "unknown";
}

function touchBucket(key: string, windowMs: number): Bucket {
  const now = Date.now();
  const current = buckets.get(key);

  if (!current || current.resetAt <= now) {
    const fresh = { count: 1, resetAt: now + windowMs };
    buckets.set(key, fresh);
    return fresh;
  }

  current.count += 1;
  return current;
}

function sweepExpiredBuckets() {
  const now = Date.now();
  for (const [key, bucket] of buckets) {
    if (bucket.resetAt <= now) {
      buckets.delete(key);
    }
  }
}

setInterval(sweepExpiredBuckets, BUCKET_SWEEP_INTERVAL_MS).unref();

export const corsMiddleware = cors({
  origin: process.env.CORS_ORIGIN ?? "*",
  allowMethods: ["GET", "OPTIONS"],
  allowHeaders: ["Content-Type", "Authorization"],
});

export const slowDownMiddleware: MiddlewareHandler = async (c, next) => {
  if (c.req.method === "OPTIONS") return next();

  if (buckets.size > BUCKET_MAX_SIZE) {
    sweepExpiredBuckets();
  }
  const key = `slow:${getClientKey(c.req.raw)}:${c.req.path}`;
  const bucket = touchBucket(key, SLOW_DOWN_WINDOW_MS);

  if (bucket.count > SLOW_DOWN_DELAY_AFTER) {
    const steps = bucket.count - SLOW_DOWN_DELAY_AFTER;
    const wait = Math.min(steps * SLOW_DOWN_DELAY_MS, SLOW_DOWN_MAX_DELAY_MS);
    await new Promise((resolve) => setTimeout(resolve, wait));
  }

  await next();
};

export const rateLimitMiddleware: MiddlewareHandler = async (c, next) => {
  if (c.req.method === "OPTIONS") return next();

  if (buckets.size > BUCKET_MAX_SIZE) {
    sweepExpiredBuckets();
  }
  const key = `limit:${getClientKey(c.req.raw)}:${c.req.path}`;
  const bucket = touchBucket(key, RATE_LIMIT_WINDOW_MS);

  const remaining = Math.max(0, RATE_LIMIT_MAX - bucket.count);
  c.header("X-RateLimit-Limit", String(RATE_LIMIT_MAX));
  c.header("X-RateLimit-Remaining", String(remaining));
  c.header("X-RateLimit-Reset", String(Math.floor(bucket.resetAt / 1000)));

  if (bucket.count > RATE_LIMIT_MAX) {
    const retryAfterSec = Math.max(1, Math.ceil((bucket.resetAt - Date.now()) / 1000));
    c.header("Retry-After", String(retryAfterSec));
    return c.json({ error: "Too many requests, please try again later." }, 429);
  }

  await next();
};

File: middleware/trafficControl.go
package middleware

import (
	"math"
	"strings"
	"sync"
	"time"

	"github.com/gofiber/fiber/v2"
	"github.com/gofiber/fiber/v2/middleware/cors"
	"pixivhono/config"
)

type Bucket struct {
	Count   int
	ResetAt int64 // unix timestamp in milliseconds
}

var (
	bucketsMu sync.Mutex
	buckets   = make(map[string]*Bucket)
)

func getClientKey(c *fiber.Ctx) string {
	forwarded := c.Get("x-forwarded-for")
	if forwarded != "" {
		parts := strings.Split(forwarded, ",")
		return strings.TrimSpace(parts[0])
	}
	realIp := c.Get("x-real-ip")
	if realIp != "" {
		return realIp
	}
	return "unknown"
}

func touchBucket(key string, windowMs int64) *Bucket {
	bucketsMu.Lock()
	defer bucketsMu.Unlock()

	now := time.Now().UnixNano() / int64(time.Millisecond)
	current, exists := buckets[key]

	if !exists || current.ResetAt <= now {
		fresh := &Bucket{Count: 1, ResetAt: now + windowMs}
		buckets[key] = fresh
		return fresh
	}

	current.Count += 1
	// Return a copy to avoid race conditions
	return &Bucket{Count: current.Count, ResetAt: current.ResetAt}
}

func sweepExpiredBuckets() {
	bucketsMu.Lock()
	defer bucketsMu.Unlock()

	now := time.Now().UnixNano() / int64(time.Millisecond)
	for k, bucket := range buckets {
		if bucket.ResetAt <= now {
			delete(buckets, k)
		}
	}
}

func StartSweepTimer() {
	interval := config.GlobalConfig.RateLimitSweepIntervalMs
	if interval <= 0 {
		interval = 30000
	}
	ticker := time.NewTicker(time.Duration(interval) * time.Millisecond)
	go func() {
		for range ticker.C {
			sweepExpiredBuckets()
		}
	}()
}

func CORSMiddleware() fiber.Handler {
	origin := config.GlobalConfig.CORSOrigin
	if origin == "" {
		origin = "*"
	}
	return cors.New(cors.Config{
		AllowOrigins: origin,
		AllowMethods: "GET,OPTIONS",
		AllowHeaders: "Content-Type,Authorization",
	})
}

func SlowDownMiddleware() fiber.Handler {
	return func(c *fiber.Ctx) error {
		if c.Method() == "OPTIONS" {
			return c.Next()
		}

		bucketsMu.Lock()
		size := len(buckets)
		bucketsMu.Unlock()
		if size > config.GlobalConfig.RateLimitBucketMaxSize {
			sweepExpiredBuckets()
		}

		key := "slow:" + getClientKey(c) + ":" + c.Path()
		bucket := touchBucket(key, int64(config.GlobalConfig.SlowDownWindowMs))

		if bucket.Count > config.GlobalConfig.SlowDownDelayAfter {
			steps := bucket.Count - config.GlobalConfig.SlowDownDelayAfter
			waitMs := int64(steps) * int64(config.GlobalConfig.SlowDownDelayMs)
			maxDelayMs := int64(config.GlobalConfig.SlowDownMaxDelayMs)
			if waitMs > maxDelayMs {
				waitMs = maxDelayMs
			}
			time.Sleep(time.Duration(waitMs) * time.Millisecond)
		}

		return c.Next()
	}
}

func RateLimitMiddleware() fiber.Handler {
	return func(c *fiber.Ctx) error {
		if c.Method() == "OPTIONS" {
			return c.Next()
		}

		bucketsMu.Lock()
		size := len(buckets)
		bucketsMu.Unlock()
		if size > config.GlobalConfig.RateLimitBucketMaxSize {
			sweepExpiredBuckets()
		}

		key := "limit:" + getClientKey(c) + ":" + c.Path()
		bucket := touchBucket(key, int64(config.GlobalConfig.RateLimitWindowMs))

		limit := config.GlobalConfig.RateLimitMax
		remaining := limit - bucket.Count
		if remaining < 0 {
			remaining = 0
		}

		c.Set("X-RateLimit-Limit", strconvItoa(limit))
		c.Set("X-RateLimit-Remaining", strconvItoa(remaining))
		c.Set("X-RateLimit-Reset", strconvItoa64(bucket.ResetAt/1000))

		if bucket.Count > limit {
			nowMs := time.Now().UnixNano() / int64(time.Millisecond)
			diffSec := float64(bucket.ResetAt-nowMs) / 1000.0
			retryAfterSec := int(math.Ceil(diffSec))
			if retryAfterSec < 1 {
				retryAfterSec = 1
			}

			c.Set("Retry-After", strconvItoa(retryAfterSec))
			return c.Status(fiber.StatusTooManyRequests).JSON(fiber.Map{
				"error": "Too many requests, please try again later.",
			})
		}

		return c.Next()
	}
}

// custom integer conversions to avoid external dependencies
func strconvItoa(val int) string {
	if val == 0 {
		return "0"
	}
	var buf [32]byte
	i := len(buf) - 1
	neg := val < 0
	if neg {
		val = -val
	}
	for val > 0 {
		buf[i] = byte('0' + (val % 10))
		i--
		val /= 10
	}
	if neg {
		buf[i] = '-'
		i--
	}
	return string(buf[i+1:])
}

func strconvItoa64(val int64) string {
	if val == 0 {
		return "0"
	}
	var buf [32]byte
	i := len(buf) - 1
	neg := val < 0
	if neg {
		val = -val
	}
	for val > 0 {
		buf[i] = byte('0' + (val % 10))
		i--
		val /= 10
	}
	if neg {
		buf[i] = '-'
		i--
	}
	return string(buf[i+1:])
}
